src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -256,182 +253,45 @@ private unsafe IAsyncResult BeginWaitForConnection(AsyncCallback callback, Objec
-                {
-                    if (cancellationHelper != null)
-                    {
-                        cancellationHelper.ThrowIOOperationAborted();
-                    }
+                    case Interop.mincore.Errors.ERROR_IO_PENDING:
+                        break;
+
+                    // If we are here then the pipe is already connected, or there was an error
+                    // so we should unpin and free the overlapped.
+                    case Interop.mincore.Errors.ERROR_PIPE_CONNECTED:
+                        // IOCompletitionCallback will not be called because we completed synchronously.
+                        completionSource.ReleaseResources();
+                        if (State == PipeState.Connected)
+                        {
+                            throw new InvalidOperationException(SR.InvalidOperation_PipeAlreadyConnected);
+                        }
+                        completionSource.SetCompletedSynchronously();
+
+                        // We return a cached task instead of TaskCompletionSource's Task allowing the GC to collect it.
+                        return Task.CompletedTask;
+
+                    default:
+                        completionSource.ReleaseResources();
+                        throw Win32Marshal.GetExceptionForWin32Error(errorCode);
                 }
-                throw Win32Marshal.GetExceptionForWin32Error(afsar._errorCode);
-            }
-
-            // Success
-            State = PipeState.Connected;
-        }
-
-        // Callback to be called by the OS when completing the async WaitForConnection operation.
-        [SecurityCritical]
-        unsafe private static void AsyncWaitForConnectionCallback(uint errorCode, uint numBytes, PipeAsyncResult asyncResult)
-        {
-            // Special case for when the client has already connected to us.
-            if (errorCode == Interop.mincore.Errors.ERROR_PIPE_CONNECTED)
-            {
-                errorCode = 0;
             }
 
-            asyncResult._errorCode = (int)errorCode;
-
-            // Call the user-provided callback.  It can and often should
-            // call EndWaitForConnection.  There's no reason to use an async 
-            // delegate here - we're already on a threadpool thread.  
-            // IAsyncResult's completedSynchronously property must return
-            // false here, saying the user callback was called on another thread.
-            asyncResult._completedSynchronously = false;
-            asyncResult._isComplete = true;
-
-            // The OS does not signal this event.  We must do it ourselves.
-            ManualResetEvent wh = asyncResult._waitHandle;
-            if (wh != null)
+            // If we are here then connection is pending.
+            // Check if the caller could cancel and if he could register cancellation.
+            if (cancellationToken.CanBeCanceled)
```

Comment:
This check isn't necessary since _cancellationToken.CanBeCanceled is checked at the beginning of the RegisterForCancellation method anyways on line 68 of ConnectionCompletionSource